### PR TITLE
Square function

### PR DESCRIPTION
**Pull request by Bigbadwolf21**
---
name: solution review 
about: A template PR for code review with a checklist
---

# Behavior

## Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [ ] A clear docstring
  - [ ] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [ ] There are tests for defensive assertions
- [ ] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [ ] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [ ] Type
  - [ ] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [ ] Each assumption about an argument is checked with an assertion
  - [ ] Each assertion checks for _only one_ assumption about the argument
- [ ] Include 3 or more (passing!) doctests

### The Function

- [ ] The function's name describes it's behavior
- [ ] The function's name matches the file name
- [ ] The function has correct type annotations

### Do's

- [ ] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in the documentation
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [ ] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [ ] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [ ] The code includes defensive assertions
- [ ] Defensive assertions include as little logic as possible

### CODE REVIEW

File: solutions/square.py [hunk 1]
```diff
@@ -0,0 +1,12 @@
+# The square root function of our calculator application
```
Comment (Allukmon) [inline solutions/square.py:1]: The comment did not align with what the function does. I think it will be better if it talks about the square.
It could be #, the function that evaluates the square of a number.

File: solutions/square.py [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+def square(num):
+    """
+    Calculate the square of a number.
+
+    Parameters:
+    number (float): The number to be squared.
```
Comment (Allukmon) [inline solutions/square.py:7]: It will be nice to make the function evaluate both the integer and float data type. 

the function could be written as 
def square(number: Union[int, float]) -> None:
    if isinstance(number, int):
        print(f"number is an integer: {number}")
    elif isinstance(number, str):
        print(f"number is a string: {number}")

File: solutions/square.py [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+
+    Returns:
+    float: The square of the input number.
```
Comment (Allukmon) [inline solutions/square.py:10]: returns should be float and int

Comment (Allukmon) [reply solutions/square.py:10]: you want to make the documentation include doctest too to make the code meet the Code review standard

File: solutions/tests/test_square.py [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+        """
+        Test cases for the square function.
+
+        Verifies the function correctly calculates the square of:
+        - Positive numbers
+        - Negative numbers
+        - Zero
+        - Floating-point numbers
+        """
+        self.assertEqual(square(2), 4)
```
Comment (Allukmon) [inline solutions/tests/test_square.py:24]: the code is correct but it will be better to make make it meet the code review standard by treating each self-assert Equal has a unit test

File: solutions/square.py (hunk 2 above)
Comment (Bigbadwolf21) [reply solutions/square.py:7]: Okay, thanks.
I will have to import Union for the code to work

Comment (Bigbadwolf21) [reply solutions/square.py:7]: The function evaluates both integer and float datatype.
I will update the docstring.

**Comment by Bigbadwolf21** [comment]
I have made all necessary corrections.
Kindly review again

File: solutions/square.py [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+    >>> square(2)
+    4
+    >>> square(-3)
+    9
+    >>> square(0)
+    0
+    >>> square(1.5)
+    2.25
+    """
+    result = num**2
```
Comment (zeozeph) [inline solutions/square.py:21]: please, also consider including defensive assertions as it is pivotal in implementing defensive checks for invalid inputs (e.g., types other than int or float) can prevent potential issues.

Example: 

def square(num: int | float) -> int | float:
    if not isinstance(num, (int, float)):
        raise TypeError(f"invalid input: {num}.The input must be an int or float")
    result = num**2
    return int(result) if isinstance(num, int) else float(result)

PS: please don't also forget to include this defensive assertions in you docstring. 

overall, you undoubtedly executed a great job!

File: solutions/tests/test_square.py [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+        self.assertAlmostEqual(square(1.5), 2.25)
+
+    def test_square_negative_float(self):
+        """
+        Test if square function correctly squares a negative float.
+        Verifies the function correctly calculates the square of:
+        - Negatve Decimal Numbers
+        """
+        self.assertAlmostEqual(square(-2.5), 6.25)
+
```
Comment (zeozeph) [inline solutions/tests/test_square.py:54]: Please consider adding tests for boundary cases, especially for extremely large and small numbers, to see if your solution can handle them or throws an error due to overflow. This will enhance your work and ensure that your work is robust enough to handle every case.

File: solutions/square.py [hunk 7]
```diff
@@ -0,0 +1,22 @@
+def square(num: int | float) -> int | float:
```
[the remark this answers is not in the document]
Comment (Bigbadwolf21) [reply solutions/square.py:15]: Resolved
